src/components/shared/StackSettings.js

```diff
@@ -266,21 +271,28 @@ export const StackSettings = explicitConnect<
   StateProps,
   DispatchProps
 >({
-  mapStateToProps: (state) => ({
-    invertCallstack: getInvertCallstack(state),
-    selectedTab: getSelectedTab(state),
-    showUserTimings: getShowUserTimings(state),
-    implementationFilter: getImplementationFilter(state),
-    currentSearchString: getCurrentSearchString(state),
-    hasJsAllocations: selectedThreadSelectors.getHasJsAllocations(state),
-    hasNativeAllocations:
-      selectedThreadSelectors.getHasNativeAllocations(state),
-    canShowRetainedMemory:
-      selectedThreadSelectors.getCanShowRetainedMemory(state),
-    callTreeSummaryStrategy:
-      selectedThreadSelectors.getCallTreeSummaryStrategy(state),
-    allowSwitchingStackType: getProfileUsesMultipleStackTypes(state),
-  }),
+  mapStateToProps: (state) => {
+    const thread = selectedThreadSelectors.getThread(state);
+    const { samples, jsAllocations, nativeAllocations } = thread;
+    return {
+      invertCallstack: getInvertCallstack(state),
+      selectedTab: getSelectedTab(state),
+      showUserTimings: getShowUserTimings(state),
+      implementationFilter: getImplementationFilter(state),
+      currentSearchString: getCurrentSearchString(state),
+      hasUsefulTimingSamples: hasUsefulSamples(samples, thread),
+      hasUsefulJsAllocations:
+        jsAllocations !== undefined && hasUsefulSamples(jsAllocations, thread),
+      hasUsefulNativeAllocations:
+        nativeAllocations !== undefined &&
+        hasUsefulSamples(nativeAllocations, thread),
```

Comment:
This works well, thanks!
But I believe there's a performance penalty here. Indeed with the `react-redux` library, it's good to remember that _all_ `mapStateToProps` functions (from all components) are called each time the state changes. Therefore it's important that it runs as fast as possible.
The problem here is that `hasUsefulSamples` might loop over the full stack table (when it's full of `null` or full of `(root)` for example, which definitely happens when profiling with `No Periodic Sampling`).

One good way to achieve this is by creating 3 memoized selectors for each of these values.

Reply:
Fixed by making the 3 memoized `src/selectors/per-thread/thread.js` and using them.